containers/src/Data/Sequence/Internal.hs

```diff
@@ -4199,6 +4201,22 @@ partition p = toPair . foldl' part (empty :*: empty)
 filter :: (a -> Bool) -> Seq a -> Seq a
 filter p = foldl' (\ xs x -> if p x then xs `snoc'` x else xs) empty
 
+-- | \( O(n) \). Take a sequence of 'Maybe's and return a sequence of all the
+-- 'Just' values.
+--
+-- @since FIXME
+catMaybes :: Seq (Maybe a) -> Seq a
+catMaybes = mapMaybe id
+
+-- | \( O(n) \). Map values and collect the 'Just' results.
+--
+-- @since FIXME
+mapMaybe :: (a -> Maybe b) -> Seq a -> Seq b
+mapMaybe f = foldl' go empty
+  where go xs x = case f x of
+          Nothing -> xs
+          Just x' -> xs `snoc'` x'
+
```

Comment:
I'm somewhat skeptical about this being the most efficient implementation. How does it compare to one that uses an intermediate list?

Reply:
No idea :sweat_smile: I don't currently know how to benchmark these things but I can look into it. I just copied `filter` so I assumed it would be fine (though I used a `where` instead of an inline function, dunno if that makes a difference).

Reply:
It is reasonable for `filter` and `mapMaybe` to have similar behavior (as I wrote in https://github.com/haskell/containers/issues/346#issuecomment-2726199830), so I think this is fine.

We can check later if both can be improved.